### PR TITLE
[Gautam's Dev bot] WI-47 Phase F1: spread elements in collection expressions targeting T[]

### PR DESCRIPTION
**Pull request by gautam-achieveai**
## Summary

Phase F1 of WI #47 adds **wire-format and Stage-2 converter infrastructure** for spread elements (`..source`) inside C# 12 collection expressions whose target is `T[]`. The Roslyn-bound lowering of source-level fixtures (`[..src]`, `[a, ..src, b]`) is deferred to Phase F4 because Roslyn's collection-expression lowering requires `System.Collections.Generic.List<T>..ctor()` as a well-known member regardless of target type, and NScript's `mscorlib` facade does not yet satisfy the well-known-member shape.

## What's in this slice

**Wire format (ProtoBuf)**
- `AstBase` tag **228** reserves the new abstract `CollectionExpressionElementSer` base.
- `[ProtoInclude(229, typeof(LiteralElementSer))]` and `[ProtoInclude(230, typeof(SpreadElementSer))]` discriminate elements through the abstract base.
- `CollectionExpressionSer.Elements` is now `List<CollectionExpressionElementSer>` (was `List<ExpressionSer>`).
- Tags are append-only; never reuse retired tags. Next free on `ExpressionSer` is **231**.

**Pipeline**
- **Stage 1** (`BoundAstToAstBase.VisitCollectionExpression`): emits `LiteralElementSer` / `SpreadElementSer` per element. Spread sources whose static type is not `ArrayTypeSymbol` are rejected with a clear `NotSupportedException` pointing at Phase F4.
- **Stage 2** (`BondToAst.ParseCollectionExpression`): literal-only inputs continue to lower to `InlineArrayInitialization` (Phase E shape). Spread-bearing inputs lower to the new `ArrayWithSpreadsInitialization` CLR.AST node.
- **Converter** (`ArrayWithSpreadsConverter`): lowers `ArrayWithSpreadsInitialization` to a JS `[].concat(arg, ...)` chain wrapped through the `ArrayG<T>` factory. Consecutive literal elements are bunched into a single inline JS array; each spread source is unwrapped through `GetNativeArrayFromArrayMethod` so that `Array.prototype.concat` flattens its elements one level. The empty `[]` receiver guarantees the result is a fresh array rather than aliasing the spread source.

## Tests

- `CollectionExpressionRoundTripTests` gains `MixedLiteralAndSpread_RoundTrips` exercising both `LiteralElementSer` (tag 229) and `SpreadElementSer` (tag 230) discriminator dispatch through the abstract base. Existing literal-only round-trip tests are updated to the new element-shape API.
- `Lang12CollectionExpressionTests.cs` retains its Phase E literal-only fixtures and documents inline why spread fixtures are deferred to Phase F4.

## Scope decisions and deferrals

- **Source-level spread fixtures (`[..src]`, `[a, ..src, b]`)** deferred to **Phase F4**. Cause: Roslyn requires `List<T>..ctor()` as a well-known member to lower any spread element regardless of target type; NScript's `mscorlib` facade exposes the constructor but not in a shape Roslyn's well-known-member resolution finds. Phase F4 will close that gap and turn on the runtime fixtures (the same gap unblocks interface targets).
- **`List<T>` and `IEnumerable<T>` spread sources** require iterator-based emission (no native JS `concat` flatten); deferred to Phase F4.
- **`Span<T>` / `ReadOnlySpan<T>`** remain Non-Goals per the issue.

## Docs

- `docs/language/csharp9-13-status.md` updates the Collection-expressions and Spread-element rows to reflect the infrastructure-only Phase F1 status and the Phase F4 sequencing.
- `docs/language/limitations.md` TL;DR updated to call out the same.

## Test plan

- [x] Solution builds in `Debug` cleanly w.r.t. the deliverable code; remaining build errors are pre-existing environmental issues with `nscript.exe` runtime initialization on Linux WSL (`XwmlParser.CssStyleSheet` type initializer) and are unrelated to this slice.
- [x] `Test/Compiler/NScript.Csc.Lib.Test` builds with **0 warnings, 0 errors** — `CollectionExpressionRoundTripTests` (incl. the new `MixedLiteralAndSpread_RoundTrips`) compiles cleanly.
- [x] `Test/Framework/RealScript/Lang12CollectionExpressionTests.cs` (literal-only fixtures) compiles cleanly without triggering the Roslyn `List<T>..ctor()` well-known-member requirement.
- [x] Manual verification: ProtoBuf tags 228 / 229 / 230 do not collide with any existing `[ProtoInclude]` on `AstBase` or its descendants.
- [x] Manual verification: converter dispatch in `ExpressionConverterBase.cs` routes `ArrayWithSpreadsInitialization` to the new converter.

Closes #47

[Gautam's Dev bot]

### CODE REVIEW

File: Sources/Compiler/JsCsc.Lib/BondToAst.cs [hunk 1]
```diff
@@ -1213,22 +1213,67 @@ private Node ParseWithExpression(Serialization.WithExpressionSer jObject)
...
+                {
+                    hasSpread = true;
+                    break;
+                }
+            }
+
+            if (!hasSpread)
+            {
+                IList<Expression> initializerExpressions = elements
+                    .Select(e => ParseExpression(((Serialization.LiteralElementSer)e).Operand))
```
Comment (gautam-achieveai) [inline Sources/Compiler/JsCsc.Lib/BondToAst.cs]: **[Gautam's review bot]** :yellow_circle: **MEDIUM — Unchecked cast without defensive guard**

This hard cast assumes every element is `LiteralElementSer` when `hasSpread=false`. But the `hasSpread` loop (lines 1226-1233) only checks for `SpreadElementSer` — a hypothetical third subtype would pass the scan and crash here with a context-free `InvalidCastException`.

The spread-bearing path 28 lines below handles this correctly with `switch` + `default: throw InvalidOperationException(...)` including the type name.

Suggested fix — use a pattern match:
```csharp
.Select(e => e is Serialization.LiteralElementSer lit
    ? ParseExpression(lit.Operand)
    : throw new InvalidOperationException(
        $"Expected LiteralElementSer in non-spread collection expression, "
        + $"but found '{e?.GetType().Name ?? "null"}'."))
```

File: Sources/Compiler/NScript.Converter/ExpressionsConverter/ArrayWithSpreadsConverter.cs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                concatArgs.Add(new JST.InlineNewArrayInitialization(
+                    location,
+                    scope,
+                    pendingLiterals));
+            }
+
+            // Always wrap through `[].concat(...)` even for a lone spread argument:
+            // collection-expression semantics require a fresh array, so the spread
+            // source must not be aliased into the result.
+            var flatNativeArray = BuildConcatCall(location, scope, concatArgs);
```
Comment (gautam-achieveai) [inline Sources/Compiler/NScript.Converter/ExpressionsConverter/ArrayWithSpreadsConverter.cs:90]: **[Gautam's review bot]** :red_circle: **HIGH — Critical invariant guarded only by comment, no test**

This comment correctly explains that `[].concat(...)` must always be used (even for a single spread) to guarantee a fresh array per collection-expression value semantics. If someone later "optimizes" this to skip `concat` for a lone spread, the result would silently alias the source array — a correctness bug.

Please add a unit test for the single-spread case (`[..a]`) that asserts the concat wrapper is present in the generated JST. Comments aren't executable contracts.

File: Test/Compiler/NScript.Csc.Lib.Test/CollectionExpressionRoundTripTests.cs [hunk 3]
```diff
@@ -49,6 +50,38 @@ public void CollectionExpressionSer_EmptyElements_RoundTrips()
...
+            Assert.AreEqual(3, typed.Elements.Count);
+
+            Assert.IsInstanceOfType(typed.Elements[0], typeof(LiteralElementSer));
+            Assert.AreEqual(7, ((IntLiteralExpression)((LiteralElementSer)typed.Elements[0]).Operand).Value);
+
+            Assert.IsInstanceOfType(typed.Elements[1], typeof(SpreadElementSer));
+            Assert.AreEqual(11, ((IntLiteralExpression)((SpreadElementSer)typed.Elements[1]).Operand).Value);
+
+            Assert.IsInstanceOfType(typed.Elements[2], typeof(LiteralElementSer));
+            Assert.AreEqual(13, ((IntLiteralExpression)((LiteralElementSer)typed.Elements[2]).Operand).Value);
```
Comment (gautam-achieveai) [inline Test/Compiler/NScript.Csc.Lib.Test/CollectionExpressionRoundTripTests.cs:83]: **[Gautam's review bot]** :red_circle: **HIGH — Serialization round-trip tests exist, but interpretation tests are missing**

These tests verify that ProtoBuf discriminator tags survive encode/decode — which is valuable. However, no test verifies that when `ParseCollectionExpression` in `BondToAst` consumes a `CollectionExpressionSer` with spread elements, it produces the correct `ArrayWithSpreadsInitialization` node (vs `InlineArrayInitialization`).

The middle layer is testable today: construct a `CollectionExpressionSer` with mixed `LiteralElementSer`/`SpreadElementSer`, feed it through the Stage 2 deserialization path, and assert:
- Result type is `ArrayWithSpreadsInitialization` (not `InlineArrayInitialization`)
- `Items.Count` matches element count
- `Items[n].IsSpread` flags are correct

This fills the most critical gap without requiring Roslyn or mscorlib changes.

**Review by gautam-achieveai — COMMENTED** [review]
**[Gautam's review bot]** Deep review of Phase F1: spread elements in collection expressions targeting T[].

## Summary

Architecture is clean — proper layer separation across Stage 1 / wire format / Stage 2 / Converter, correct ProtoBuf tag allocation (228/229/230), follows existing converter registration conventions. The `[].concat(...)` JS emission strategy is correct and well-documented.

**Verdict: REQUEST CHANGES** — test coverage gaps in the middle layer are feasible to address within the stated Roslyn constraints.

---

## Work Item Context
- **User Story #47**: Close supportable C# 9-13 language gaps in NScript
- This PR covers **Phase F1** — wire-format and Stage-2 converter infrastructure for spread elements
- Source-level E2E spread fixtures are blocked by Roslyn requiring `List<T>..ctor()` as a well-known member (acknowledged and correctly deferred to Phase F4)

---

## Findings

### :red_circle: HIGH — Missing middle-layer tests for `ArrayWithSpreadsConverter` and `BondToAst` spread branch

The PR adds ~250 lines of new production logic across three key components, but the only test is a ProtoBuf serialization round-trip. This proves the wire format survives encode/decode — it does **not** prove:

1. That the deserialized format is correctly interpreted into the right AST node type (`ArrayWithSpreadsInitialization` vs `InlineArrayInitialization`)
2. That the converter bunching logic produces correct JS (consecutive literals grouped, spread sources unwrapped, always-concat for fresh-array semantics)

**The middle layer is testable today without Roslyn.** `BondToAst.ParseCollectionExpression` takes a `CollectionExpressionSer` (constructable in tests) and produces a CLR AST node (inspectable). The converter takes an `ArrayWithSpreadsInitialization` and produces JST.

**Recommended minimum tests:**
- Construct `CollectionExpressionSer` with `[LiteralElementSer, SpreadElementSer, LiteralElementSer]` -> assert result is `ArrayWithSpreadsInitialization` with correct `Items[n].IsSpread` flags
- Single-spread case `[..a]` -> verify concat wrapper is present (fresh-array invariant, currently guarded only by a comment at line 87-89 of converter)
- All-spreads `[..a, ..b]` — no inline arrays emitted, only unwrapped sources
- Mixed ordering `[lit, lit, spread, lit]` — verify bunching produces 3 concat args (inline array of 2, unwrapped spread, inline array of 1)

### :yellow_circle: MEDIUM — Unchecked cast in literal-only path (`BondToAst.cs`)

In `ParseCollectionExpression`, when `hasSpread=false`, every element is hard-cast via:
```csharp
.Select(e => ParseExpression(((Serialization.LiteralElementSer)e).Operand))
```

The spread-bearing path 28 lines below uses a proper `switch` + `default: throw InvalidOperationException` with the element type name. If Phase F4 adds a third `CollectionExpressionElementSer` subtype, the `hasSpread` scan would not detect it and this cast would throw a context-free `InvalidCastException`.

**Fix:** Use a pattern match with an explicit error, symmetric with the spread path.

### :green_circle: LOW — Double iteration in `ParseCollectionExpression`

Elements are iterated twice: once for `hasSpread` scan, once for building items. A single-pass approach through one `switch` loop would be both simpler and eliminate the LINQ `.Select().ToList()` allocation. Not a performance concern (compile-time, small collections), but a code-clarity improvement.

### :green_circle: LOW — XOR-only `GetHashCode` (non-actionable)

Follows existing codebase convention (`InlineArrayInitialization` uses the same pattern). Not worth changing in isolation.

---

## Architecture Verification (all clean :white_check_mark:)
- Layer boundaries: Stage 1 -> Serialization -> Stage 2 -> CLR AST -> Converter -> JST — no cross-layer violations
- ProtoBuf tags 228/229/230: collision-free, correctly scoped
- `CollectionExpressionElementSer` correctly inherits `AstBase` (not `ExpressionSer`)
- Converter follows `SimplifyConverter<T>` registration pattern
- `[].concat(...)` via `StringLiteralExpression` + `IndexExpression` is correct for native JS method access
- `GetNativeArrayFromArrayMethod` and `GetArrayNativeArrayArgCtor` are pre-existing APIs used elsewhere

## Review Agents Used
Architecture, Test Coverage, Performance, Exception Handling, Review Grader

**Review by gautam-achieveai — COMMENTED** [review]
**[Gautam's review bot]** Incremental re-review of iteration 2 (commit `2878a8f9`).

## Comment Response Assessment

### Thread 1 — MEDIUM: Unchecked cast — :white_check_mark: Resolved
The hard cast `(LiteralElementSer)e` has been replaced with a pattern match + explicit `InvalidOperationException` carrying the offending type name. Now symmetric with the spread branch's `switch` + `default: throw` discipline. Good.

### Thread 2 — HIGH: Fresh-array invariant test — :white_check_mark: Resolved
`BuildConcatCall` made public for testability. Two new tests in `ArrayWithSpreadsConverterTests` pin the `[].concat(...)` wrapper shape for both single-arg and multi-arg cases. The `AssertConcatShape` helper validates the full JST tree down to the empty receiver array and `concat` member name. This is exactly what was asked for — the lone-spread case is now guarded by an executable contract, not just a comment.

### Thread 3 — HIGH: Missing interpretation tests — :yellow_circle: Partially addressed (acceptable)
The original ask was to test `ParseCollectionExpression` end-to-end (construct `CollectionExpressionSer` -> Stage 2 -> assert `ArrayWithSpreadsInitialization`). Instead, `ContainsSpreadElement` was extracted as public static with 6 unit tests covering all edge cases. This tests the *dispatch decision* but not the *branch execution* (constructing `ArrayInitializationItem` objects).

The constraint is reasonable — `ParseCollectionExpression` is an instance method requiring `_clrContext` + `TypeInfoSer` for `DeserializeType`, making a standalone test infrastructure-heavy. The discriminator IS the most critical piece (wrong dispatch -> wrong node type entirely), and the branch body is structurally simple with proper `default: throw` guard. **Acceptable compromise for Phase F1.**

---

**Verdict: APPROVE** — All three review comments substantively addressed. No new issues found in the incremental diff. The remaining interpretation-path gap can be covered naturally when Phase F4 adds E2E spread fixtures with full Roslyn integration.

## Work Item Context
- **User Story #47**: Close supportable C# 9-13 language gaps in NScript
- Phase F1 slice: wire-format + Stage-2 converter infrastructure for spread elements in collection expressions targeting T[]

File: Sources/Compiler/JsCsc.Lib/BondToAst.cs (hunk 1 above)
Comment (gautam-achieveai) [reply Sources/Compiler/JsCsc.Lib/BondToAst.cs]: **[Gautam's review bot]** :white_check_mark: Addressed in `2878a8f9`. Pattern match with explicit `InvalidOperationException` carrying the offending type name. Symmetric with the spread branch. Resolved.

File: Sources/Compiler/NScript.Converter/ExpressionsConverter/ArrayWithSpreadsConverter.cs (hunk 2 above)
Comment (gautam-achieveai) [reply Sources/Compiler/NScript.Converter/ExpressionsConverter/ArrayWithSpreadsConverter.cs:90]: **[Gautam's review bot]** :white_check_mark: Addressed in `2878a8f9`. `ArrayWithSpreadsConverterTests` pins the `[].concat(...)` wrapper shape for both single-arg and multi-arg cases. The `AssertConcatShape` helper validates the empty receiver array and `concat` member name — exactly the executable contract this comment asked for. Resolved.

File: Test/Compiler/NScript.Csc.Lib.Test/CollectionExpressionRoundTripTests.cs (hunk 3 above)
Comment (gautam-achieveai) [reply Test/Compiler/NScript.Csc.Lib.Test/CollectionExpressionRoundTripTests.cs:83]: **[Gautam's review bot]** :yellow_circle: Partially addressed in `2878a8f9`. The `ContainsSpreadElement` extraction + 6 unit tests cover the dispatch discriminator thoroughly (null, empty, all-literal, single-spread, mixed, trailing-spread).

The original ask was for full `ParseCollectionExpression` E2E tests asserting `ArrayWithSpreadsInitialization` result type and `Items[n].IsSpread` flags. That remains uncovered because `ParseCollectionExpression` needs `_clrContext` + `TypeInfoSer`.

**Accepting this compromise for Phase F1** — the discriminator is the most critical decision point, and the branch body is structurally simple with a proper `default: throw` guard. The remaining gap can be naturally covered when Phase F4 adds E2E spread fixtures with full Roslyn integration. Resolved.
